### PR TITLE
2) Per-process power policy

### PR DESCRIPTION
**Pull request by MatejKafka**
Currently, the policy and the config frequency parameters are not documented; I'll add the documentation when the interface is finalized. Other than that, this is ready for a review.

TODO: Figure out a good interface to set frequencies for the processes without explicitly-defined frequencies; currently, the last explicitly-set frequency is kept. This is efficient, as there aren't needless switches, but it is not very intuitive and all frequency changes must be set manually. Another possible solution could be to have a single fixed frequency for all other processes, or use the original idea of allowing an additional power policy, which only applies to the processes without an explicitly-set frequency.

### CODE REVIEW

File: src/config.cpp [hunk 1]
```diff
@@ -302,6 +289,30 @@ Node Config::normalize_window(const Node &win,  // in: window to normalize
     return norm_win;
 }
 
+static unsigned int imx8_freq_i_from_freq(bool is_a53, size_t freq_MHz)
```
Comment (wentasah) [inline src/config.cpp:292]: I'm not happy that the information about available frequencies needs to be duplicated here. I wonder whether it's necessary to have this function and `Config::validate_per_process_freq_feasibility`. Wouldn't it be simpler to validate the feasibility later, e.g. in `DemosSched` constructor or `setup()` or at least in `config.create_scheduler_objects()`. At this points the `PowerPolicy` is already available and checking of power policy feasibility can be done better. Perhaps, the `PowerPolicy` class can have `validate()` method which would implement all the checks. Would there be any problem with that?

**Review by wentasah — COMMENTED** [review]
See my comments in-line.

File: src/config.cpp (hunk 1 above)
Comment (MatejKafka) [reply src/config.cpp:292]: Overall, the reason I implemented the checks in `Config` is to allow to user to statically validate the config without executing it. I agree it's not nice in the code, but the UX is better this way. :) However, it only works while we support only i.MX8.

The conversion from frequency in MHz to the frequency index is not necessary, but it's easier to work with, as I can use a bitmap for the AND/OR operations during the validation. It can be done using `std::set`, but it's a lot less efficient and the API is not very nice. I could create the bitmap dynamically before knowing the frequencies by creating a `std::map` `frequency => bitmap index` dynamically as I encounter the frequencies in config. That way `Config` doesn't need to know the exact frequencies and there's no duplication.

However, I still need to know the layout of the CPU clusters, which I cannot get outside `PowerPolicy`, and it depends on the current machine. Either I'll duplicate it in `Config` and it will be i.MX8-specific, or I can move the validation to `PowerPolicy` and make it dependent on the current machine. As I noted in the `Config::validate_per_process_freq_feasibility` docstring, it is possible to verify the feasibility later, and it would work for any CPU (not just i.MX8). However, the check wouldn't run during config parsing.

To move the validation to the power policy, it would need to know about either the `Config` instance, or all `Process` and `Window` instances. It is imo OK (design-wise) to pass them to a method like `void PowerPolicy::validate_configuration(Config c)`.


TL;DR: The validation can be kept in `Config` and improved to not depend on knowing the frequencies ahead-of-time, or it can be moved to PowerPolicy, where the check occurs later, depends on current machine, but it's more flexible and works everywhere (not just i.MX8). The first one is less work, the second one is probably a bit cleaner.

**Comment by wentasah** [comment]
On Sat, Aug 07 2021, Matej Kafka wrote:
> TL;DR: The validation can be kept in `Config` and improved to not
> depend on knowing the frequencies ahead-of-time, or it can be moved to
> PowerPolicy, where the check occurs later, depends on current machine,
> but it's more flexible and works everywhere (not just i.MX8). The
> first one is less work, the second one is probably a bit cleaner.

I just don't want to add extra high complexity for a feature that is
maybe just an experimental power policy. I have the impression that if
we check for problematic configurations at run time rather than ahead of
time, the implementation would be significantly simpler. If this is be
the case, I'd be completely fine with issuing an error at run time.

In general we assume that major frame lengths are bellow 1 second, and I
think it's acceptable if one has to wait 1 second for an error message.

File: src/partition.hpp [hunk 2]
```diff
@@ -53,6 +53,8 @@ class Partition
     // Note: `processes` MUST be located after all cgroups (cgc, cgf, cge) - Process destructors
     //  must run before the cgroup destructors to cleanup child cgroups in correct order
     Processes processes{};
+    /** Current cpu_set the partition is running under. */
+    cpu_set current_cpus{};
```
[the remark this answers is not in the document]
Comment (MatejKafka) [reply src/partition.hpp:57]: It seems weird to me to ask the cgroup class what the current cpuset is, when the process sets it.

**Comment by MatejKafka** [comment]
I've moved the checks to runtime, as refactoring the code to support the ahead-of-time checks without duplication would be quite complex and require coupling between `PowerPolicy`and `Config`. When we finalize this feature, it shouldn't be too hard to move the checks back to ahead-of-time if deemed necessary.

Currently, if the runtime check detects a collision, an exception is thrown, which stops DEmOS without cleanup. Fixing this would require storing the exception, doing a cleanup asynchronously and then rethrowing it. I can implement the fix in 2-3 hours if necessary, but given that it's an edge condition that shouldn't normally happen, I don't think it's a high priority issue right now.

File: src/partition.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (wentasah) [reply src/partition.hpp]: See newly added commit 632652b for what I meant.

**Comment by wentasah** [comment]
This is quite hard for me to review. The addition of code and its removal in subsequent commits makes it difficult to review it commit by commit and modification of 22 files in total makes it difficult to review it as a whole. Nevertheless, it looks more or less good. I'll be playing with it and doing some changes myself. I'll let you know later.

File: src/partition.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (MatejKafka) [reply src/partition.hpp]: I understand that; it seemed to me that conceptually, the current cpuset is a property of the partition, not the cgroup, but it's true that this way, there's less duplication. :+1:

**Comment by wentasah** [comment]
No, I already have some changes on top of this.

**Comment by wentasah** [comment]
I did a few changes - mainly:
- I removed imx8-specific naming of config keys (if we will ever need to support migration between clusters, we'll allow the frequency value be an array - see e06fc02ccfca5ff016d1d1bec59df289013a0fdd commit message for details).
- I added ahead-of-time validation of configured frequencies.
- I added a new policy called imx8_per_slice (we'll need this for some experiments).

I'm going to merge this now, but I'd be happy if @MatejKafka looks at this and points out potential problems etc.

**Comment by MatejKafka** [comment]
I probably could have communicated this a bit better – you've re-implemented some parts of the original ahead-of-time solution I had, and I had a few architectural reasons to avoid exactly this change to the PowerPolicy class. :)

I have to work on the other parts of my assignment now, but with these changes, it should be quite simple to use the ahead-of-time validation instead of the runtime detection.

**Comment by wentasah** [comment]
Ok. DEmOS can currently do exactly what we need for our experiments, so it's fine to work on something else :-)
